src/server/pfs/server/commit_set.go

```diff
@@ -0,0 +1,450 @@
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+	totalCommits := len(commitInfos)
+	result := make([]*pfs.CommitInfo, 0)
+	collected := make(map[string]struct{})
+	collectCount := 0
+	// O(n^2) sorting of commits
+	//
+	// TODO(acohen4) FIXME: it's possible for a commit to have two branches from the same repo in its provenance.
+	// this must be keyed by branch
+	// COULD WE instead sort by created time? DOES CREATED TIME COMPLETELY IMPLY TOPOLOGICAL ORDERING?
+	for collectCount < totalCommits {
+		for i, ci := range commitInfos {
+			satisfied := true
+			if _, ok := collected[pfsdb.RepoKey(ci.Commit.Repo)]; !ok {
+				for _, p := range ci.DirectProvenance {
+					if _, ok := collected[pfsdb.RepoKey(p.Repo)]; !ok {
+						satisfied = false
+						break
+					}
+				}
+			}
+			if satisfied {
+				collectCount++
+				collected[pfsdb.RepoKey(ci.Commit.Repo)] = struct{}{}
+				if !filterPassiveCommits || ci.Commit.ID == commitset.ID {
+					result = append(result, ci)
+				}
+				commitInfos[i] = commitInfos[len(commitInfos)-1]
+				commitInfos = commitInfos[:len(commitInfos)-1]
+				break
+			}
+		}
+	}
+	return result, nil
+}
+
+func (d *driver) inspectCommitSetImmediate(ctx context.Context, commitset *pfs.CommitSet, cb func(*pfs.CommitInfo) error) error {
+	var commitInfos []*pfs.CommitInfo
+	if err := d.txnEnv.WithReadContext(ctx, func(txnCtx *txncontext.TransactionContext) error {
+		var err error
+		commitInfos, err = d.inspectCommitSetImmediateTx(txnCtx, commitset, false)
+		return err
+	}); err != nil {
+		return err
+	}
+	for _, commitInfo := range commitInfos {
+		if err := cb(commitInfo); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// applies a callback, cb, to all the commits in a commit set. A commit set includes
+func (d *driver) inspectCommitSet(ctx context.Context, commitset *pfs.CommitSet, wait bool, cb func(*pfs.CommitInfo) error) error {
+	if !wait {
+		return d.inspectCommitSetImmediate(ctx, commitset, cb)
+	}
+	sent := map[string]struct{}{}
+	send := func(ci *pfs.CommitInfo) error {
+		if _, ok := sent[pfsdb.CommitKey(ci.Commit)]; ok {
+			return nil
+		}
+		sent[pfsdb.CommitKey(ci.Commit)] = struct{}{}
+		return cb(ci)
+
+	}
+	unfinishedCommits := make([]*pfs.Commit, 0)
+	// NOTE: before 2.5, a triggered commits would be included as part of the same commit set as the triggering commit set,
+	// so we would wait for all of the current commit set to finish, then check if new previously unknown commits are added due to triggers.
+	if err := d.inspectCommitSetImmediate(ctx, commitset, func(ci *pfs.CommitInfo) error {
+		if ci.Finished != nil {
+			return send(ci)
+		}
+		unfinishedCommits = append(unfinishedCommits, proto.Clone(ci.Commit).(*pfs.Commit))
+		return nil
+	}); err != nil {
+		return err
+	}
+	for _, uc := range unfinishedCommits {
+		// TODO: make a dedicated call just for the blocking part, inspectCommit is a little heavyweight?
```

Comment:
Undone TODO

Reply:
This is just a copy and paste TODO. I think it is fine as is for this PR.